### PR TITLE
Add visitor to remove bias from bn_ layer inputs (#closes 2155)

### PR DESCRIPTION
**Pull request by arrufat**
As I explained in #2155, this PR adds a visitor that will automatically remove bias learning from all layers that are inputs of `bn_` layers.

### CODE REVIEW

**Comment by arrufat** [comment]
I think it's ready for review :)

File: dlib/dnn/layers.h [hunk 1]
```diff
@@ -1522,6 +1522,36 @@ namespace dlib
...
+            void set_input_no_bias(T&) const
+            {
+                // ignore other layer types
+            }
+
+            template <layer_mode mode, typename U, typename E>
+            void set_input_no_bias(add_layer<bn_<mode>, U, E>& l)
+            {
+                l.subnet().layer_details().set_bias_learning_rate_multiplier(0);
+                l.subnet().layer_details().set_bias_weight_decay_multiplier(0);
```
Comment (davisking) [inline dlib/dnn/layers.h:1540]: This should be this instead right?
```
        set_bias_learning_rate_multiplier(l.subnet().layer_details(), 0);
        set_bias_weight_decay_multiplier(l.subnet().layer_details(), 0);
```
just in case someone sticks a batch norm layer on top of something that doesn't have a `.set_bias_learning_rate_multiplier`

File: dlib/dnn/layers.h [hunk 2]
```diff
@@ -1533,6 +1563,14 @@ namespace dlib
...
     }
 
+    template <typename net_type>
+    void set_all_bn_inputs_no_bias (
+        net_type& net
+    )
+    {
+        visit_layers(net, impl::visitor_bn_input_no_bias());
+    }
+
```
Comment (davisking) [inline dlib/dnn/layers.h:1573]: This should probably go in core.h next to the other visitors that do this kind of thing.  Like right below `set_bias_learning_rate_multiplier` is reasonable.

Comment (arrufat) [reply dlib/dnn/layers.h:1573]: I wanted to put it there in the first place, but I just put it here because the template parameter is `layer_mode` is defined in this file. I guess that's also why you put here the `visitor_bn_running_stats_window_size` and the `set_all_bn_running_stats_window_sizes`.

File: dlib/dnn/layers.h (hunk 1 above)
Comment (arrufat) [reply dlib/dnn/layers.h:1540]: Oh, good catch! thank you :)

**Comment by arrufat** [comment]
> Thanks for the PR :)
> 
> As an side, you might want to add a `.disable_bias()` to the layers that makes them not execute the bias addition at all, since that will run faster. Then instead of calling `set_bias_learning_rate_multiplier()` you could call an equivalent `disable_bias()`.

Yes, that was bothering me, the fact that even all the values are 0, we're still performing the computation. I stared looking at the `fc_no_bias` to see how you disabled the bias there. I will try to do it :)

File: dlib/dnn/layers.h (hunk 2 above)
Comment (davisking) [reply dlib/dnn/layers.h:1573]: Oh right, I forgot about `set_all_bn_running_stats_window_sizes`. Yeah, putting it here seems good :+1:

**Comment by davisking** [comment]
> > Thanks for the PR :)
> > As an side, you might want to add a `.disable_bias()` to the layers that makes them not execute the bias addition at all, since that will run faster. Then instead of calling `set_bias_learning_rate_multiplier()` you could call an equivalent `disable_bias()`.
> 
> Yes, that was bothering me, the fact that even all the values are 0, we're still performing the computation. I stared looking at the `fc_no_bias` to see how you disabled the bias there. I will try to do it :)

Na, don't do it like `fc_no_bias`.  It's probably nicer and simpler if this is just a normal runtime `bool` in each applicable layer that controls if the bias is added or not. Then a member function can just set it to true or false.

**Comment by arrufat** [comment]
Done, I hope this is what you expected. I had to modify the serialize functions as well, but I made sure to keep backwards compatibility: I've successfully loaded and run the following models:
-  resnet34 model, which contains `fc` layers
- dlib_face_recognition_resnet_model, which contains `fc_no_bias` and `con` layers
- a saved sync file from the dcgan example, which contains `con` and `cont` layers.

I've also checked that the DCGAN example still runs. I've modified a little bit to print out the number of trainable parameters, to ensure the bias are not allocated when disabling the bias.

File: dlib/dnn/layers.h [hunk 3]
```diff
@@ -1522,6 +1564,37 @@ namespace dlib
...
+            {
+                // ignore other layer types
+            }
+
+            template <layer_mode mode, typename U, typename E>
+            void set_input_no_bias(add_layer<bn_<mode>, U, E>& l)
+            {
+                disable_bias(l.subnet().layer_details());
+                set_bias_learning_rate_multiplier(l.subnet().layer_details(), 0);
+                set_bias_weight_decay_multiplier(l.subnet().layer_details(), 0);
```
Comment (arrufat) [inline dlib/dnn/layers.h:1583]: I know these two functions are redundant, but I thought it was good to keep them any way.

File: dlib/dnn/layers.h [hunk 4]
```diff
@@ -289,8 +305,11 @@ namespace dlib
                 << ")";
             out << " learning_rate_mult="<<item.learning_rate_multiplier;
             out << " weight_decay_mult="<<item.weight_decay_multiplier;
-            out << " bias_learning_rate_mult="<<item.bias_learning_rate_multiplier;
-            out << " bias_weight_decay_mult="<<item.bias_weight_decay_multiplier;
+            if (item.use_bias)
+            {
+                out << " bias_learning_rate_mult="<<item.bias_learning_rate_multiplier;
+                out << " bias_weight_decay_mult="<<item.bias_weight_decay_multiplier;
+            }
```
Comment (davisking) [inline dlib/dnn/layers.h:312]: `to_xml()` needs to be updated as well.  There adding a `use_bias=true` or `use_bias=false` attribute is reasonable.

File: dlib/dnn/layers.h (hunk 3 above)
Comment (davisking) [reply dlib/dnn/layers.h:1583]: Yeah, seems reasonable.

File: dlib/dnn/layers_abstract.h [hunk 5]
```diff
@@ -573,6 +573,13 @@ namespace dlib
                 - #get_bias_weight_decay_multiplier() == val
         !*/
 
+        void disable_bias(
+        );
+        /*!
+            ensures
+                - disables bias for this layer
```
Comment (davisking) [inline dlib/dnn/layers_abstract.h:580]: Add a `bool bias_is_disabled()` method too so that someone can tell if a layer has had its bias disabled or not.  Then make the docs for the `disable_bias()` methods note that they set `bias_is_disabled()` to false.

**Comment by arrufat** [comment]
Do you think I need update the `operator<<` so that I show the bias status? Right now, the only way to know is by looking at whether the bias learning rates are printed.

**Comment by davisking** [comment]
> Do you think I need update the `operator<<` so that I show the bias status? Right now, the only way to know is by looking at whether the bias learning rates are printed.

I thought about that earlier and I'm on the fence.  `operator<<` output is just for humans to eyeball and at some level adding more and more stuff to it makes it busy and hard to read.  Take a look at it for a printout of a network and set it however you think looks the most clear in that context.  But likely what you have in there now is fine.

**Comment by arrufat** [comment]
I think it's more clear this way:
- if we are using bias, keep it as it was, that is, print the bias multipliers
- otherwise, do not print those, and print `use_bias=false`, so that users don't think there has been an error, or the output has been truncated.

I think it's a good trade-off, since we are not making the lines with bias even longer, and the ones without are still short in comparison.

**Comment by davisking** [comment]
> I think it's more clear this way:
> 
> * if we are using bias, keep it as it was, that is, print the bias multipliers
> * otherwise, do not print those, and print `use_bias=false`, so that users don't think there has been an error, or the output has been truncated.
> 
> I think it's a good trade-off, since we are not making the lines with bias even longer, and the ones without are still short in comparison.

Sounds good :)

**Comment by davisking** [comment]
Sweet, thanks for another cool PR :)

**Comment by arrufat** [comment]
Thank you! I am really learning a lot by doing these things :)
